File: compare_routes.py
import re, json
from pathlib import Path

base = Path(r"c:/Users/InformaTeam/Desktop/b2b Project/e-comerce/backend")
routes_file = base / "routes" / "api.php"
postman_file = base / "postman" / "B2B-Mediation-API.postman_collection.json"

def get_routes():
    if not routes_file.exists(): return []
    text = routes_file.read_text(encoding="utf-8")
    lines = text.splitlines()
    prefix_stack = [""]
    routes = []
    for line in lines:
        s = line.strip()
        m = re.search(r"Route::prefix\(['\"]([^'\"]+)['\"]\)->group\(function", s)
        if m:
            cur = prefix_stack[-1]
            new = (cur + "/" + m.group(1)).replace("//", "/")
            prefix_stack.append(new)
            continue
        if s.startswith("});"):
            if len(prefix_stack) > 1: prefix_stack.pop()
            continue
        m2 = re.search(r"Route::(get|post|put|patch|delete)\s*\(\s*['\"]([^'\"]+)['\"]", s, re.I)
        if m2:
            method = m2.group(1).upper()
            path = m2.group(2)
            prefix = prefix_stack[-1]
            full = (prefix + "/" + path).replace("//", "/")
            routes.append((method, full))
        m3 = re.search(r"Route::apiResource\(\s*['\"]([^'\"]+)['\"]", s)
        if m3:
            resource = m3.group(1)
            prefix = prefix_stack[-1]
            basep = (prefix + "/" + resource).replace("//", "/")
            routes.extend([
                ("GET", basep), ("POST", basep),
                ("GET", basep + "/{id}"), ("PATCH", basep + "/{id}"), ("DELETE", basep + "/{id}")
            ])
    return routes

def get_postman():
    if not postman_file.exists(): return []
    with open(postman_file, "r", encoding="utf-8") as f:
        data = json.load(f)
    endpoints = []
    def walk(items):
        for it in items:
            if "request" in it:
                req = it["request"]
                method = req.get("method", "GET").upper()
                url = req.get("url")
                path = ""
                if isinstance(url, dict):
                    if "path" in url and isinstance(url["path"], list):
                        path = "/" + "/".join([str(p) for p in url["path"] if p])
                    elif "raw" in url:
                        path = url["raw"]
                elif isinstance(url, str):
                    path = url
                endpoints.append((method, path))
            if "item" in it:
                walk(it["item"])
    walk(data.get("item", []))
    return endpoints

def normalize(path):
    path = re.sub(r"https?://[^/]+", "", path)
    path = path.split("?")[0]
    path = re.sub(r"\{\{.*?\}\}", "{id}", path)
    path = re.sub(r"\{.*?\}", "{id}", path)
    path = re.sub(r"/\d+", "/{id}", path)
    path = re.sub(r"/+", "/", path).strip("/")
    return path

laravel_routes = get_routes()
postman_endpoints = get_postman()
missing = []
norm_pm = set((m, normalize(p)) for m, p in postman_endpoints)
for m, p in sorted(set(laravel_routes)):
    if (m, normalize(p)) not in norm_pm:
        missing.append((m, p))

print(json.dumps({"laravel_count": len(set(laravel_routes)), "postman_count": len(postman_endpoints), "missing": missing}, indent=2))

File: compare_routes.php
$routesFile = $argv[1];
$postmanFile = $argv[2];
$routes = file_get_contents($routesFile);
$lines = preg_split("/\r?\n/", $routes);
$prefixStack = [""];
$routesArr = [];
foreach ($lines as $line) {
    $s = trim($line);
    if (preg_match("/Route::prefix\([''']([^''']+)[''']\)->group\(function\(\) \{/", $s, $m)) {
        $cur = end($prefixStack);
        $new = rtrim($cur, "/") . "/" . trim($m[1], "/");
        $prefixStack[] = preg_replace("#/+#", "/", $new);
        continue;
    }
    if ($s == "});") {
        if (count($prefixStack) > 1) array_pop($prefixStack);
        continue;
    }
    if (preg_match("/Route::(get|post|put|patch|delete)\s*\(\s*[''' ]([^''' ]+)[''' ]/", $s, $m2)) {
        $method = strtoupper($m2[1]);
        $path = $m2[2];
        $prefix = end($prefixStack);
        $full = preg_replace("#/+#", "/", $prefix . "/" . $path);
        if (substr($full, 0, 1) != "/") $full = "/" . $full;
        $routesArr[] = [$method, $full];
        continue;
    }
    if (preg_match("/Route::apiResource\(\s*[''' ]([^''' ]+)[''' ]/", $s, $m3)) {
        $resource = $m3[1];
        $prefix = end($prefixStack);
        $basepath = preg_replace("#/+#", "/", $prefix . "/" . $resource);
        if (substr($basepath, 0, 1) != "/") $basepath = "/" . $basepath;
        $routesArr[] = ["GET", $basepath];
        $routesArr[] = ["POST", $basepath];
        $routesArr[] = ["GET", $basepath . "/{id}"];
        $routesArr[] = ["PATCH", $basepath . "/{id}"];
        $routesArr[] = ["DELETE", $basepath . "/{id}"];
        continue;
    }
}
$normRoutes = [];
foreach ($routesArr as $r) {
    list($m, $p) = $r;
    $p = preg_replace("#/+#", "/", $p);
    $normRoutes[$m . " " . $p] = 1;
}
$pm = json_decode(file_get_contents($postmanFile), true);
$pmSet = [];
function walk($items, &$pmSet) {
    if (!is_array($items)) return;
    foreach ($items as $it) {
        if (isset($it["request"])) {
            $req = $it["request"];
            $method = strtoupper($req["method"] ?? "GET");
            $url = $req["url"];
            $path = "";
            if (isset($url["path"]) && is_array($url["path"])) {
                $path = "/" . implode("/", $url["path"]);
            } elseif (isset($url["raw"])) {
                $raw = $url["raw"];
                $path = parse_url($raw, PHP_URL_PATH);
                if (!$path) {
                    if (strpos($raw, "{{") !== false) {
                        $p = preg_replace("/^https?:\/\/[^\/]+/", "", $raw);
                        $p = explode("?", $p)[0];
                        $path = $p;
                    } else {
                        $path = $raw;
                    }
                }
            }
            $path = preg_replace("#/+#", "/", "/".$path);
            $pmSet[$method . " " . $path] = 1;
        }
        if (isset($it["item"])) walk($it["item"], $pmSet);
    }
}
walk($pm["item"], $pmSet);
$missing = [];
foreach (array_keys($normRoutes) as $k) {
    $found = false;
    list($rMethod, $rPath) = explode(" ", $k, 2);
    $rNorm = rtrim(preg_replace("#\{[^/]+\}#", "{id}", $rPath), "/");
    foreach (array_keys($pmSet) as $pk) {
        list($pmMethod, $pmPath) = explode(" ", $pk, 2);
        if ($pmMethod !== $rMethod) continue;
        $pmNorm = preg_replace("#/\d+#", "/{id}", $pmPath);
        $pmNorm = rtrim(preg_replace("#\{[^/]+\}#", "{id}", $pmNorm), "/");
        if ($pmNorm === $rNorm) {
            $found = true;
            break;
        }
    }
    if (!$found) $missing[] = $k;
}
echo json_encode(["routes_count" => count($normRoutes), "postman_count" => count($pmSet), "missing" => $missing]);
